src/analysis/inner_product_space/gram_schmidt_ortho.lean

```diff
@@ -0,0 +1,323 @@
+/-
+Copyright (c) 2022 Jiale Miao. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jiale Miao, Kevin Buzzard
+-/
+
+import analysis.inner_product_space.projection
+
+/-!
+# Gram-Schmidt Orthogonalization and Orthonormalization
+
+In this file we introduce Gram-Schmidt Orthogonalization and Orthonormalization.
+
+The Gram-Schmidt process takes a set of vectors as input
+and outputs a set of orthogonal vectors which have the same span.
+
+## Main results
+
+- `gram_schmidt` : Gram-Schmidt process
+- `gram_schmidt_orthogonal` :
+  the proof that `gram_schmidt` produces an orthogonal system of vectors
+- `span_gram_schmidt` :
+  Gram-Schmidt process preserves span of vectors
+- `gram_schmidt_ne_zero` :
+  If the input of first n vectors of gram_schmidt are linearly independent
+  , then output of first n vectors are non-zero
+- `gram_schmidt_normed` :
+  Normalized "Gram-Schmidt" (i.e each vector in this system has unit length)
+- `gram_schmidt_orthornormal` :
+  the proof that `gram_schmidt_normed` produces an orthornormal system of vectors
+
+## TODO
+  Construct a version with an orthonormal basis from the Gram-Schmidt process.
+-/
+
+open_locale big_operators
+
+variables (𝕜 : Type*) {E : Type*} [is_R_or_C 𝕜] [inner_product_space 𝕜 E]
+
+local notation `⟪`x`, `y`⟫` := @inner 𝕜 _ _ x y
+
+/-- Gram-Schmidt process -/
+noncomputable def gram_schmidt (f : ℕ → E) : ℕ → E
+| n := f n - ∑ i : fin n, orthogonal_projection (𝕜 ∙ gram_schmidt i) (f n)
+using_well_founded {dec_tac := `[exact i.prop]}
+
+/-- 'gram_schmidt_def' turns the sum over `fin n` into a sum over `ℕ`. -/
+lemma gram_schmidt_def (f : ℕ → E) (n : ℕ) :
+  gram_schmidt 𝕜 f n = f n - ∑ i in finset.range n,
+    orthogonal_projection (𝕜 ∙ gram_schmidt 𝕜 f i) (f n) :=
+begin
+  rw gram_schmidt,
+  congr' 1,
+  exact fin.sum_univ_eq_sum_range (λ i,
+    (orthogonal_projection (𝕜 ∙ gram_schmidt 𝕜 f i) (f n) : E)) n,
+end
+
+@[simp] lemma gram_schmidt_zero (f : ℕ → E) :
+  gram_schmidt 𝕜 f 0 = f 0 :=
+by simp only [gram_schmidt, fintype.univ_of_is_empty, finset.sum_empty, sub_zero]
+
+/-- Gram-Schmidt process produces an orthogonal system of vectors. -/
+theorem gram_schmidt_orthogonal' (f : ℕ → E) (a b : ℕ) (h₀ : a < b) :
+  ⟪gram_schmidt 𝕜 f a, gram_schmidt 𝕜 f b⟫ = 0 :=
+begin
```

Comment:
This is a good use case for `suffices`:
```suggestion
theorem gram_schmidt_orthogonal (f : ℕ → E) (a b : ℕ) (h₀ : a ≠ b) :
  ⟪gram_schmidt 𝕜 f a, gram_schmidt 𝕜 f b⟫ = 0 :=
begin
  suffices : ∀ a b : ℕ, a < b → ⟪gram_schmidt 𝕜 f a, gram_schmidt 𝕜 f b⟫ = 0,
  { cases h₀.lt_or_lt with ha hb,
    { exact this _ _ ha, },
    { rw inner_eq_zero_sym,
      exact this _ _ hb, }, },
  clear h₀ a b,
  intros a b h₀,
```
then you can delete the lemma below

Reply:
oh yes, that becomes more clear now, I have changed it, thank you!